File: app/src/main/java/deltazero/amarok/utils/LauncherIconController.java
package deltazero.amarok.utils;

import android.app.Activity;
import android.content.ComponentName;
import android.content.pm.PackageManager;

import deltazero.amarok.BuildConfig;

public class LauncherIconController {
    private static final String LAUNCHER_DEFAULT = "deltazero.amarok.launcher.default";
    private static final String LAUNCHER_CALENDAR = "deltazero.amarok.launcher.calendar";

    public enum IconState {
        VISIBLE,    // Normal Amarok icon visible
        DISGUISED,  // Calendar icon visible (disguised)
        HIDDEN      // No icon visible
    }

    public static void setIconState(Activity activity, IconState state) {
        PackageManager pm = activity.getPackageManager();

        switch (state) {
            case VISIBLE -> {
                // Show normal Amarok icon
                setComponentState(pm, LAUNCHER_DEFAULT, true);
                setComponentState(pm, LAUNCHER_CALENDAR, false);
            }
            case DISGUISED -> {
                // Show calendar icon (disguised)
                setComponentState(pm, LAUNCHER_CALENDAR, true);
                setComponentState(pm, LAUNCHER_DEFAULT, false);
            }
            case HIDDEN -> {
                // Hide all icons
                setComponentState(pm, LAUNCHER_DEFAULT, false);
                setComponentState(pm, LAUNCHER_CALENDAR, false);
            }
        }
    }

    private static void setComponentState(PackageManager pm, String componentName, boolean enabled) {
        pm.setComponentEnabledSetting(
                new ComponentName(BuildConfig.APPLICATION_ID, componentName),
                enabled ? PackageManager.COMPONENT_ENABLED_STATE_ENABLED : PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP);
    }
}

File: app/src/main/java/deltazero/amarok/utils/LauncherIconController.kt
package deltazero.amarok.utils

import android.app.Activity
import android.content.ComponentName
import android.content.pm.PackageManager
import deltazero.amarok.BuildConfig

object LauncherIconController {
  private const val LAUNCHER_DEFAULT = "deltazero.amarok.launcher.default"
  private const val LAUNCHER_CALENDAR = "deltazero.amarok.launcher.calendar"

  enum class IconState {
    VISIBLE, // Normal Amarok icon visible
    DISGUISED, // Calendar icon visible (disguised)
    HIDDEN, // No icon visible
  }

  @JvmStatic
  fun setIconState(activity: Activity, state: IconState) {
    val pm = activity.packageManager

    when (state) {
      IconState.VISIBLE -> {
        // Show normal Amarok icon
        setComponentState(pm, LAUNCHER_DEFAULT, true)
        setComponentState(pm, LAUNCHER_CALENDAR, false)
      }
      IconState.DISGUISED -> {
        // Show calendar icon (disguised)
        setComponentState(pm, LAUNCHER_CALENDAR, true)
        setComponentState(pm, LAUNCHER_DEFAULT, false)
      }
      IconState.HIDDEN -> {
        // Hide all icons
        setComponentState(pm, LAUNCHER_DEFAULT, false)
        setComponentState(pm, LAUNCHER_CALENDAR, false)
      }
    }
  }

  private fun setComponentState(pm: PackageManager, componentName: String, enabled: Boolean) {
    pm.setComponentEnabledSetting(
      ComponentName(BuildConfig.APPLICATION_ID, componentName),
      if (enabled) {
        PackageManager.COMPONENT_ENABLED_STATE_ENABLED
      } else {
        PackageManager.COMPONENT_ENABLED_STATE_DISABLED
      },
      PackageManager.DONT_KILL_APP,
    )
  }
}
